docs/contributing/javascript_guidelines.rst

```diff
@@ -1,51 +1,179 @@
-   everywhere else.
 
+Linting
+~~~~~~~
 
-Linting and formatting code
-~~~~~~~~~~~~~~~~~~~~~~~~~~~
+All JavaScript code in the Wagtail codebase is compliant with the specs from the ESLint config. After making a Pull
+Request the linter will also check all the JavaScript code, so make sure you checked you changes with ESLint before
+you commit your code.
 
-Wagtail provides some tooling configuration to help check your code meets the
-styleguide. You'll need node.js and npm on your development machine.
-Ensure project dependencies are installed by running ``npm install``
+**Running the linter**
 
+Execute ``npm run lint`` to check if your JavaScript code meets the styleguide.
 
-**Linting code**
 
-``npm run lint:js``
+Wagtail React guidelines
+~~~~~~~~~~~~~~~~~~~~~~~~
 
-This will lint all the JS in the wagtail project, excluding vendor
-files and compiled libraries.
+In addition to the JavaScript used in the ``wagtailadmin`` app, we use `React <https://facebook.github.io/react/>`_ to enhance the UI with interactive
+components, like the Explorer.
 
-Some of the modals are generated via server-side scripts. These include
-template tags that upset the linter, so modal workflow JavaScript is
-excluded from the linter.
+**ES6**
 
+In Wagtail we use `Babel <https://babeljs.io/docs/learn-es2015/>`_ with the ``es2015`` and ``react`` presets.
+Especially with React in a combination with Redux we use the following features a lot.
 
-**Formatting code**
+**Arrow functions**
 
-``npm run format:js``
+For anonymous functions you should use the arrow function feature.
 
-This will perform safe edits to conform your JS code to the styleguide.
-It won't touch the line-length, or convert quotemarks from double to single.
+.. code-block:: javascript
 
-Run the linter after you've formatted the code to see what manual fixes
-you need to make to the codebase.
+    // Bad
+    export const loginSuccess = createAction('LOGIN_SUCCESS', function(email, profile) {
+      return { email: email, profile: profile };
+    });
 
-**Changing the linter configuration**
+    // Good
+    export const loginSuccess = createAction('LOGIN_SUCCESS', (email, profile) => {
+      return { email, profile };
+    });
 
-Under the hood, the tasks use the `JavaScript Code Style <http://jscs.info/>`_ library.
+Note that we always wrap single arguments with brackets.
 
-To edit the settings for ignored files, or to change the linting rules,
-edit the ``.jscsrc`` file in the wagtail project root.
+**Spread operators (instead of Object.assign)**
 
-A complete list of the possible linting rules can be found here:
-`JSCS Rules <http://jscs.info/rules.html>`_
+.. code-block:: javascript
+
+    // Bad
+    case 'LOGIN_SUCCESS':
+      return Object.assign({}, state, {
+        email: action.payload.email,
+        profile: action.payload.profile
+      });
+
+    // Good
+    case 'LOGIN_SUCCESS':
+      return { ...state, email: action.payload.email, profile: action.payload.profile };
```

Comment:
Is this ES6 or ES6+? I thought this was https://github.com/sebmarkbage/ecmascript-rest-spread, which is still in stage 2 (https://github.com/tc39/proposals). Do we need to use the stage 2 preset?